validations/src/validations.rs

```diff
@@ -416,161 +355,203 @@ pub fn validate_transaction_signature(
         }
     }
 
+    // Validate reveal transactions in a block
+    let mut re_mt = ProgressiveMerkleTree::sha256();
+    let mut reveals_number = HashMap::new();
+    for transaction in &block.txns.reveal_txns {
+        let (dr_pointer, dr_witnesses, fee) = validate_reveal_transaction(&transaction, dr_pool)?;
+
+        increment_witnesses_counter(
+            &mut reveals_number,
+            &dr_pointer,
+            u32::from(dr_witnesses),
+            fee,
+        );
+
+        // Add new hash to merkle tree
+        let txn_hash = transaction.hash();
+        let Hash::SHA256(sha) = txn_hash;
+        re_mt.push(Sha256(sha));
+    }
+    let re_hash_merkle_root = re_mt.root();
+
+    // Add reveal fees
+    for WitnessesCount { fee, .. } in reveals_number.values() {
+        total_fee += fee;
+    }
+
+    // Validate tally transactions in a block
+    let mut ta_mt = ProgressiveMerkleTree::sha256();
+    for transaction in &block.txns.tally_txns {
+        let (outputs, fee) = validate_tally_transaction(transaction, dr_pool)?;
+        total_fee += fee;
+
+        update_utxo_diff(&mut utxo_diff, vec![], outputs, transaction.hash());
+
+        // Add new hash to merkle tree
+        let txn_hash = transaction.hash();
+        let Hash::SHA256(sha) = txn_hash;
+        ta_mt.push(Sha256(sha));
+    }
+    let ta_hash_merkle_root = ta_mt.root();
+
     // Validate mint
     validate_mint_transaction(
-        mint,
+        &block.txns.mint,
         total_fee,
         block_reward(block.block_header.beacon.checkpoint),
     )?;
+    let mint_hash_merkle_root = hash_merkle_tree_root(&[block.txns.mint.hash()]);
 
     // Insert mint in utxo
-    let mint_output_pointer = OutputPointer {
-        transaction_id: mint.hash(),
-        output_index: 0,
-    };
-    utxo_diff.insert_utxo(mint_output_pointer, mint.outputs[0].clone());
-
-    // Validate commits number
-    for WitnessesCount { current, target } in commits_number.values() {
-        if current != target {
-            Err(BlockError::MismatchingCommitsNumber {
-                commits: *current,
-                rf: *target,
-            })?
-        }
-    }
+    update_utxo_diff(
+        &mut utxo_diff,
+        vec![],
+        block.txns.mint.outputs.iter().collect(),
+        block.txns.mint.hash(),
+    );
 
     // Validate Merkle Root
-    let Hash::SHA256(mr) = block.block_header.hash_merkle_root;
-    if mt.root() != Sha256(mr) {
+    let Hash::SHA256(vt_mr) = block.block_header.merkle_roots.vt_hash_merkle_root;
+    let Hash::SHA256(dr_mr) = block.block_header.merkle_roots.dr_hash_merkle_root;
+    let Hash::SHA256(co_mr) = block.block_header.merkle_roots.commit_hash_merkle_root;
+    let Hash::SHA256(re_mr) = block.block_header.merkle_roots.reveal_hash_merkle_root;
+    let Hash::SHA256(ta_mr) = block.block_header.merkle_roots.tally_hash_merkle_root;
+
+    if (mint_hash_merkle_root != block.block_header.merkle_roots.mint_hash_merkle_root)
```

Comment:
You could create a new `BlockMerkleRoots` and just compare it with `block.block_header.merkle_roots`